File: Round 2 2011/B.cpp
// Problem: Spinning Blade
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <iostream>
#include <string>

using namespace std;

int mb[500][500][3];
int m[3];
int s[500][500][3];
int g[500][500];

string compute()
{
    int r, c, d;
    cin >> r >> c >> d;
    for (int i = 0; i < r; i++)
    {
	for (int j = 0; j < c; j++)
	{
	    char ch;
	    cin >> ch;
	    g[i][j] = (int)ch - 48;
	}
    }
    for (int i = 0; i < r; i++)
    {
	for (int j = 0; j < c; j++)
	{
	    int mi = g[i][j] + d;
	    mb[i][j][0] = mi;
	    mb[i][j][1] = i*mi;
	    mb[i][j][2] = j*mi;
	}
    }
    for (int k = 0; k < 3; k++)
    {
	for (int i = 0; i < r; i++)
	{
	    for (int j = 0; j < c; j++)
	    {
		if (i == 0 && j == 0)
		    s[i][j][k] = mb[i][j][k];
		else if (i == 0)
		    s[i][j][k] = s[i][j-1][k] + mb[i][j][k];
		else if (j == 0)
		    s[i][j][k] = s[i-1][j][k] + mb[i][j][k];
		else
		    s[i][j][k] = s[i-1][j][k] + s[i][j-1][k]
			         - s[i-1][j-1][k] + mb[i][j][k];
	    }
	}
    }
    int tm = -1;
    for (int r1 = 0; r1 < r; r1++)
    {
	for (int c1 = 0; c1 < c; c1++)
	{
	    for (int r2 = r1+2; r2 < r; r2++)
	    {
		if (c1 + (r2-r1) < c)
		{
		    int c2 = c1 + r2 - r1;
		    for (int k = 0; k < 3; k++)
		    {
			m[k] = s[r2][c2][k];
			if (r1 >= 1)
			{
			    m[k] -= s[r1-1][c2][k];
			    if (c1 >= 1)
			    {
				m[k] += s[r1-1][c1-1][k];
			    }
			}
			if (c1 >= 1)
			{
			    m[k] -= s[r2][c1-1][k];
			}
			
			m[k] -= mb[r1][c1][k] + mb[r1][c2][k]
			        + mb[r2][c1][k] + mb[r2][c2][k];
			if (m[0] > 0 && (r1+r2)*m[0] == 2*m[1]
			             && (c1+c2)*m[0] == 2*m[2])
			{
			    if (r2-r1+1 > tm)
			    {
				tm = r2-r1+1;
			    }
			}
		    }
		}
	    }
	}
    }
    if (tm < 0)
    {
	cout << "IMPOSSIBLE" << endl;
    }
    else
    {
	cout << tm << endl;
    }
}

int main()
{
    int cases;
    cin >> cases;

    for (int i = 1; i <= cases; i++)
    {
	cout << "Case #" << i << ": ";
	compute();
    }
    
    return 0;
}

File: Round 2 2011/B.py
#!/usr/bin/python
#
# Problem: Spinning Blade
# Language: Python
# Author: KirarinSnow
# Usage: pypy thisfile.py <input.in >output.out 
# Comments: Runs in 7 minutes using pypy for the large input.


mb = []
for i in range(500):
    mb.append([])
    for j in range(500):
        mb[i].append(None)
s = []
for i in range(500):
    s.append([])
    for j in range(500):
        s[i].append(None)

def compute():
    r, c, d = map(int, raw_input().split())
    g = [map(int, raw_input()) for i in range(r)]
    for i in range(r):
        for j in range(c):
            mi = g[i][j]+d
            mb[i][j] = (mi, i*mi, j*mi)

    for i in range(r):
        for j in range(c):
            if i == 0 and j == 0:
                s[i][j] = mb[i][j]
            elif i == 0:
                s[i][j] = map(lambda k: s[i][j-1][k] + mb[i][j][k], range(3))
            elif j == 0:
                s[i][j] = map(lambda k: s[i-1][j][k] + mb[i][j][k], range(3))
            else:
                s[i][j] = map(lambda k: s[i-1][j][k] + s[i][j-1][k] -
                              s[i-1][j-1][k] + mb[i][j][k], range(3))

    tm = -1
    for r1 in range(r):
        for c1 in range(c):
            for r2 in range(r1+2, r):
                if c1 + (r2-r1) < c:
                    c2 = c1 + (r2-r1)
                    
                    m = [0]*3
                    for k in range(3):
                        m[k] = s[r2][c2][k]
                        if r1-1 >= 0:
                            m[k] -= s[r1-1][c2][k]
                            if c1-1 >= 0:
                                m[k] += s[r1-1][c1-1][k]
                        if c1-1 >= 0:
                            m[k] -= s[r2][c1-1][k]
                        m[k] -= mb[r1][c1][k] + mb[r1][c2][k] + \
                            mb[r2][c1][k] + mb[r2][c2][k]
                    
                    if m[0] >= 0 and (r1+r2)*m[0] == 2*m[1] and \
                                     (c1+c2)*m[0] == 2*m[2]:
                        tm = max(tm, r2-r1+1)
                        
    return "IMPOSSIBLE" if tm < 0 else tm
        
for i in range(input()):
    print "Case #%d: %s" % (i+1, compute())
